### ISSUE TITLE
Drop-down menu with many items is slow

### ISSUE THREAD

**Issue by Aquarian-Age** [fyne-io/fyne#2164]
### Describe the bug:

Select the drop-down menu year, the corresponding time is too long

### Screenshots:

![Peek 2021-04-14 06-44](https://user-images.githubusercontent.com/36611556/114630457-3bd61480-9ced-11eb-8d54-a520517b41f1.gif)

### Example code:

```go
package main

import (
        "fmt"
        "time"

        "fyne.io/fyne/v2"
        "fyne.io/fyne/v2/app"
        "fyne.io/fyne/v2/container"
        "fyne.io/fyne/v2/layout"
        "fyne.io/fyne/v2/theme"
        "fyne.io/fyne/v2/widget"
)

var t = time.Now().Local()

func main() {
        a := app.New()
        w := a.NewWindow("hello")
        a.Settings().SetTheme(theme.LightTheme())

        //select options
        years := func() []string {
                var s []string
                for i := 1601; i <= 3499; i++ {
                        is := fmt.Sprintf("%dyear", i)
                        s = append(s, is)
                }
                return s
        }

        selecty := widget.NewSelect(years(), func(s string) {
        })
        selecty.PlaceHolder = "year"

        btnOK := widget.NewButton("OK", func() {
                if selecty.SelectedIndex() == -1 {
                        selecty.SetSelectedIndex(t.Year() - 1601)
                }

                fmt.Printf("%s year index=%d\n", selecty.Selected, selecty.SelectedIndex())
        })

        hbox := container.New(layout.NewHBoxLayout(), selecty, btnOK)

        w.SetContent(container.New(layout.NewVBoxLayout(), hbox))
        w.Resize(fyne.Size{600, 600})
        w.ShowAndRun()

}
```

### Device
 - **OS: Linux** 
 - **Version: VERSION_ID="20210401" PRETTY_NAME="openSUSE Tumbleweed"**
 - **Go version: go1.16.2 linux/amd64**
 - **Fyne version: v2.0.2**

**Comment by Jacalz** [comment]
Thanks for opening this bug report. Could you please fill out the description and add notes on how to reproduce. I would also appreciate if you could trim down the code example to be abort and concise. It makes it a lot easier for us to triage and review all the issues that we get. Thank you 🙂

**Comment by Aquarian-Age** [comment]
> Thanks for opening this bug report. Could you please fill out the description and add notes on how to reproduce. I would also appreciate if you could trim down the code example to be abort and concise. It makes it a lot easier for us to triage and review all the issues that we get. Thank you

---

Select the drop-down menu year page will lose response
Select other drop-down menus, the page reacts normally

**Comment by Jacalz** [comment]
So you think that you could update the original issue description with the information and changes that I requested?

**Comment by Aquarian-Age** [comment]
> Do you think that you could update the original issue description with the information and changes that I requested?

Yes it has been updated

**Comment by andydotxyz** [comment]
Thanks, this is an issue we should resolve.
However I think that having 1898 options in a drop-down menu is not likely to be a good user experience.

**Comment by Aquarian-Age** [comment]
> Thanks, this is an issue we should resolve.
> However I think that having 1898 options in a drop-down menu is not likely to be a good user experience.

---
It’s not very good, but there are default values in the code.
```
                 if selecty.SelectedIndex() == -1 {
                         selecty.SetSelectedIndex(t.Year()-1601)
                 }
```

**Comment by Jacalz** [comment]
> > Thanks, this is an issue we should resolve.
> > However I think that having 1898 options in a drop-down menu is not likely to be a good user experience.
> 
> It’s not very good, but there are default values in the code.
> 
> ```
>                  if selecty.SelectedIndex() == -1 {
>                          selecty.SetSelectedIndex(t.Year()-1601)
>                  }
> ```

I think @andydotxyz meant it from a usability standpoint. Finding the correct value from 1898 values is never going to be a good user experience, even if you select a default. I could see pretty large amounts being useful for selecting countries or languages, but 1898 is absurdly large :)

Either way, I think I might have a little solution for this. Will see if I can get it done later in the week.

**Comment by Jacalz** [comment]
I opened up https://github.com/fyne-io/fyne/pull/2180. It is not the entire fix that I had in mind, but should make things a lot better anyhow. As I noted in the PR description, the menu-items are recreated each time the popup is shown. The next step would probably be to optimize away the need to do so, but this should at least make things better in the meantime.

**Comment by Jacalz** [comment]
I threw together a quick test for using `widget.List` for showing the popups. This works around the whole issue with having to create items for all of the options that we have. It is not working currently, but I think it should outright solve this when it finally works https://github.com/Jacalz/fyne/commit/27d5f625669a1de195e9048e6199b67a9df10708.

**Comment by Jacalz** [comment]
Some news regarding this. The PR https://github.com/fyne-io/fyne/pull/2181 is now almost entirely finished. I have tested with 10000 items and there is no delay at all. It is working buttery smooth with many items.
![fast-select](https://user-images.githubusercontent.com/25466657/115269358-85828d00-a13b-11eb-9ab9-d01f4429c0c0.gif)

**Comment by Aquarian-Age** [comment]
@Jacalz @andydotxyz Which branch should I switch to to use? Switch to git switch -c v2.0.3-rc1 branch, too many drop-down menu problems still exist

**Comment by Jacalz** [comment]
Only a minor performance optimization landed in 2.0.3. Just making things a little bit better.

The entire fix that should make large amounts of options usable is in https://github.com/fyne-io/fyne/pull/2181. It's on a branch in my fork for the moment.

**Comment by Jacalz** [comment]
It looks like I missed to convey the information that we decided a while back (quite a long time now actually) that there was a better solution to this issue than the one that I proposed in the PR. We agreed that it would be better to optimize the drop down to use similar logic as the list widget internally, instead of trying to adapt a list widget for `widget.Select` and `widget.SelectEntry` that would look like and behave like a menu. This means that all use cases where a menu issues will be seeing similar performance benefits.

I don't think that I'll have the time to dig into fixing this before 2.1 unfortunately. If anyone want to take this over and work on getting it implemented, feel free to do so.

**Comment by andydotxyz** [comment]
Yes, let's see if we can make the cache-reuse from List/Table/Tree mechanic more generic for the 2.1 release so we can use it here and also the RichText code.

**Comment by Jacalz** [comment]
@Aquarian-Age Why did you close this? It has not been fixed yet.
